src/checkbox/checkbox.js

```diff
@@ -15,255 +15,148 @@
+const classes_ = Object.freeze({
+  ROOT: 'mdl-checkbox',
+  JS: 'mdl-js-checkbox',
+  INPUT: 'mdl-checkbox__input',
+
+  IS_FOCUSED: 'is-focused',
+  IS_DISABLED: 'is-disabled',
+  IS_CHECKED: 'is-checked',
+  IS_UPGRADED: 'is-upgraded'
+});
 
+/**
+ * The MaterialCheckbox class wraps a Material Design checkbox component.
+ *
+ * @export
+ */
+class MaterialCheckbox {
   /**
-   * Class constructor for Checkbox MDL component.
-   * Implements MDL component design pattern defined at:
-   * https://github.com/jasonmayes/mdl-component-design-pattern
+   * Initialize checkbox from a DOM node.
    *
-   * @constructor
-   * @param {HTMLElement} element The element that will be upgraded.
+   * @param {Element} root The element being upgraded.
    */
-  var MaterialCheckbox = function MaterialCheckbox(element) {
-    this.element_ = element;
+  constructor(root) {
+    // Check if the root has the right class.
+    if (!root.classList.contains(classes_.ROOT)) {
+      throw new Error(`MaterialCheckbox missing ${classes_.ROOT} class.`);
+    }
+    this.root_ = root;
 
-    // Initialize instance.
-    this.init();
-  };
-  window['MaterialCheckbox'] = MaterialCheckbox;
+    // Look for required sub-nodes in the root's DOM.
+    this.input_ = root.querySelector(`.${classes_.INPUT}`);
+    if (!this.input_) {
+      throw new Error(`MaterialCheckbox missing ${classes_.INPUT} node.`);
+    }
 
-  /**
-   * Store constants in one place so they can be updated easily.
-   *
-   * @enum {string | number}
-   * @private
-   */
-  MaterialCheckbox.prototype.Constant_ = {
-    TINY_TIMEOUT: 0.001
-  };
+    // Initialize events.
+    this.input_.addEventListener('change', this.refresh.bind(this));
+    this.input_.addEventListener('focus',
+        () => this.root_.classList.add(classes_.IS_FOCUSED));
+    this.input_.addEventListener('blur',
+        () => this.root_.classList.remove(classes_.IS_FOCUSED));
+    this.root_.addEventListener('mouseup', this.blur_.bind(this));
 
-  /**
-   * Store strings for class names defined by this component that are used in
-   * JavaScript. This allows us to simply change it in one place should we
-   * decide to modify at a later date.
-   *
-   * @enum {string}
-   * @private
-   */
-  MaterialCheckbox.prototype.CssClasses_ = {
-    INPUT: 'mdl-checkbox__input',
-    BOX_OUTLINE: 'mdl-checkbox__box-outline',
-    FOCUS_HELPER: 'mdl-checkbox__focus-helper',
-    TICK_OUTLINE: 'mdl-checkbox__tick-outline',
-    RIPPLE_EFFECT: 'mdl-js-ripple-effect',
-    RIPPLE_IGNORE_EVENTS: 'mdl-js-ripple-effect--ignore-events',
-    RIPPLE_CONTAINER: 'mdl-checkbox__ripple-container',
-    RIPPLE_CENTER: 'mdl-ripple--center',
-    RIPPLE: 'mdl-ripple',
-    IS_FOCUSED: 'is-focused',
-    IS_DISABLED: 'is-disabled',
-    IS_CHECKED: 'is-checked',
-    IS_UPGRADED: 'is-upgraded'
-  };
+    // Refresh component.
+    this.refresh();
+
+    // Attach component to DOM property.
+    this.root_['MaterialCheckbox'] = this;
```

Comment:
This should be down in the loop for creating self-managed resources.

By default, developers should store references however they wish and we shouldn't add extra things to nodes. Then people that opt-in to a managed handling state get these properties added for convenience.

So line 161 here should look like `nodes[i]['MaterialCheckbox'] = new MaterialCheckbox(nodes[i]);`. And for automation later, the classAsString should be a property of the classes to be pulled from.